Project development setup

KlaasH opened the pull request:
## Overview

Sets up some of the standard project infrastructure:
- Scripts To Rule Them All: Just dev-related scripts so far (setup, update, server, lint, and format)
- Prettier:
   - Added to frontend dev dependencies
   - Project-level `.prettierrc` file
   - Runs in "check" mode in `lint` and "write" mode in `format`
   - I ran it and committed the changes it made
- README: The frontend had a standard Next.js README already, which I revised a bit but didn't delete. I added a top-level README with the standard setup instructions and list of scripts and ports.
- Dev server port: Picked a port (3008), recorded it in [the team Ports list](https://github.com/azavea/geospatial-apps/blob/master/project_ports.md), and changed all references to the default one (3000).
- Pull request template

Resolves #1

### Demo

![image](https://user-images.githubusercontent.com/6598836/155161518-31f30ff3-6fc6-45d8-9b7d-023e69072e31.png)

### Notes

It's possible we won't benefit in the long term from having the front-end in a subdirectory, in which case we might decide that setting it up that way and having to deal with that nuance (e.g. have all the STRTA change directory before they do anything) is mildly annoying on balance.  But if we do end up with a few different components here, which seems likely given that we're planning to make the data processing independent from the front-end, I think it'll be nice to not have things mixed together.

## Testing Instructions

- Run `./scripts/setup` (which actually just calls `update`)
- Run `./scripts/update` (it won't do much, since you just ran it)
- Run `./scripts/server`. Go to http://localhost:3008/ and you should see a sample graph and a links to a few other preliminary snippets.
- Run `./scripts/lint`. It should run successfully and not find any errors.
- Peruse the READMEs for missing or erroneous info, or typos.

## Checklist

- [x] `./scripts/lint` runs without errors

KlaasH committed:
Add nvm and yvm config to standardize node and yarn versions

Also deletes package-lock.json, since standardizing on yarn means not
supporting npm.

KlaasH committed:
Add setup, update, and server scripts

Adds the basic "run the project scripts". The two slightly unusual things
are that 'update' and 'dev' both active nvm and yvm and both change to the
'frontend' directory to run.
The 'setup' script just calls 'update', since there's currently nothing special
required for initial project setup.

KlaasH committed:
Add 'prettier' and apply linting and formatting changes

- Adds 'prettier' to the dev dependencies and adds a .prettierrc file
  (I put it at the project root and removed the empty one in the frontend
  directory. Might not matter, but this way if we end up with multiple
  subprojects with javascript code, they'll share a format config.)
- Adds a `lint` script to `scripts/` that runs ESLint and also runs
  `prettier` in "check" mode
- Adds a `format` script that runs prettier in 'write' mode to
  automatically fix anything it considers an error
- Applies all the changes requested by the linter or made by prettier

KlaasH committed:
Add top-level readme and edit frontend readme

Adds a top-level README with the basics--getting started, server port, STRTA.
I left the existing Next.js boilerplate README in the `frontend/` directory,
but edited it a little.

KlaasH committed:
Change dev server port to 3008

Everything on 3000 is a bad time. Picking a project port is preferable.

KlaasH committed:
Add pull request template

ddohler commented on scripts/activate_yvm:
```diff
@@ -0,0 +1,25 @@
+if [ -f "$nvm_script" ]; then
+    source "$nvm_script"
+fi
+if ! nvm list | grep -q "$(cat .nvmrc)"; then
```
This is fine as-is, but you could also handle this similarly to the tests for existence of `yvm` and `nvm`: Optimistically run `nvm use` and then install if the error code is non-zero (and then run `nvm use` again). That would leave you with two `nvm use` calls but would avoid the need to parse the output of `nvm list`.

KlaasH replied:
Ok, yeah, this seems to work:
```
if ! nvm use; then
    # If just running 'nvm use' didn't work, run 'nvm install' then try again
    nvm install "$(cat .nvmrc)"
    nvm use
fi
```
I'm not sure which I like better.  The `list | grep` way seems more elaborate but this seems a little more indirect.  I think I'll go with this, though, since it's also slightly faster.

ddohler commented on scripts/update:
```diff
@@ -0,0 +1,21 @@
+        source "$(dirname $0)"/activate_yvm
+        pushd frontend
+        echo "Updating dependencies"
+        yarn install
```
There seems to be [some debate about this](https://github.com/yarnpkg/yarn/issues/4147), but I think we've determined that `yarn install` should include `--frozen-lockfile` by default in the past: https://github.com/azavea/civic-apps-app-template/issues/82

KlaasH replied:
Hmm.  The case seems convincing, though it's interesting that 3.5 years after that issue was written, neither the civic apps template nor the react/redux/typescript one uses that option.

ddohler reviewed (approved):
Everything looks good! I had a couple small suggestions but nothing critical.

Not for this issue, but I do have a couple questions about the CI setup for this project; I'm not seeing a CI / deployment setup issue yet so I'm leaving them here, but I'm also happy to convert these into an issue if that would be preferable:
- I see that both Netlify and Vercel are enabled on this project. Are we planning to continue using both? What's the difference between Vercel and Netlify? I've never used Vercel but a (very) quick read of their site makes it sound like it does fairly similar things to how we mostly use Netlify. I'm curious to hear more about it and what the use cases are for each.
- What Netlify account is this project associated with? I tried logging in with the company account and didn't see this site listed, and I didn't see a separate account for this project in LastPass.

KlaasH committed:
Optimistically call 'nvm use' rather than checking the list first

This seems a little less direct but also simpler than the 'list | grep'
method. For a tie-breaker, it's also slightly faster.

KlaasH committed:
Use --frozen-lockfile option for 'yarn install'

KlaasH commented:
Re CI: I'm not sure either.  Possibly @jfrankl set them up under personal accounts. I think sticking with just Netlify would make sense, since it's what we use elsewhere, and that a separate issue to bring the Netlify config into the fold would be useful.

jfrankl commented:
@KlaasH Yep, this is set up on my personal Netlify. Feel free to set it up on the Azavea Netlify, or I could transfer the project. Let me know if you need my help.

jfrankl commented:
Also, I was checking out Vercel, but I agree that we should use Netlify.

ddohler commented:
Thanks for clarifying! I went ahead and opened #6 to cover removing Vercel and switching to the Azavea Netlify account.
